### ISSUE TITLE
Using MySQL instead of MariaDB

### ISSUE THREAD

**Issue by aayush27894** [bitnami/bitnami-docker-matomo#83]
I am using the following docker commands for setting up Matomo with MySQL instead of MariaDB.

```
docker network create matomo_network

docker run -d --name matomo-db -e ALLOW_EMPTY_PASSWORD=yes -e MYSQL_USER=bn_matomo -e MYSQL_DATABASE=bitnami_matomo -e MYSQL_PASSWORD=Pass@123 --net matomo_network --volume ~/mysql-d/bitnami/mysql/data bitnami/mysql:latest

docker run -d --name matomo -p 84:8080 -e ALLOW_EMPTY_PASSWORD=yes -e MATOMO_DATABASE_HOST=mysql -e MATOMO_DATABASE_USER=bn_matomo -e MATOMO_DATABASE_PASSWORD=Pass@123 -e MATOMO_DATABASE_NAME=bitnami_matomo --net matomo_network bitnami/matomo:latest
```

These are the docker logs for Matomo:

```
matomo 21:54:35.44 
matomo 21:54:35.45 Welcome to the Bitnami matomo container
matomo 21:54:35.45 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-matomo
matomo 21:54:35.45 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-matomo/issues
matomo 21:54:35.46 
matomo 21:54:35.46 INFO  ==> ** Starting Matomo setup **
matomo 21:54:35.50 INFO  ==> Configuring PHP options
matomo 21:54:35.51 INFO  ==> Validating settings in MYSQL_CLIENT_* env vars
matomo 21:54:35.55 WARN  ==> You set the environment variable ALLOW_EMPTY_PASSWORD=yes. For safety reasons, do not use this flag in a production environment.
matomo 21:54:35.61 INFO  ==> Ensuring Matomo directories exist
matomo 21:54:35.62 INFO  ==> Trying to connect to the database server
matomo 21:57:17.76 ERROR ==> Could not connect to the database
```

Could you please tell me what the docker-compose.yml file would look like if I wanted to use mysql database instead?

How would one connect to and remote MySQL database with the Bitnami Matomo container?

**Comment by marcosbc** [comment]
Hi @aayush27894, the first issue is that you're specifying a wrong host:

```diff
-docker run -d --name matomo -p 84:8080 -e ALLOW_EMPTY_PASSWORD=yes -e MATOMO_DATABASE_HOST=mysql ...
+docker run -d --name matomo -p 84:8080 -e ALLOW_EMPTY_PASSWORD=yes -e MATOMO_DATABASE_HOST=matomo-db ...
```

And then, you're running MySQL 8.0 which is not fully compatible with Matomo unless you also specify `mysql_native_password` as the MySQL authentication plugin, e.g. by also passing a new environment variable like: `-e MYSQL_AUTHENTICATION_PLUGIN=mysql_native_password`.

After that it should work fine.

**Comment by aayush27894** [comment]
Thank you for the prompt reply. 
I am now trying to connect to a MySQL (ver. 8.0.23) database that I am hosting on my local machine. I am the using a docker-compose.yml for bringing up the a matomo instance. This is what the docker compose file looks like:

```
version: '2'
services:
  matomo:
    image: "docker.io/bitnami/matomo:4-debian-10"
    ports:
      - "84:8080"
    environment:
      - MARIADB_HOST={host-machine-ip}
      - MARIADB_PORT_NUMBER=3306
      - MARIADB_ROOT_USER={local-mysql-root-user}
      - MARIADB_ROOT_PASSWORD={local-mysql-root-password}
      - MYSQL_CLIENT_CREATE_DATABASE_NAME=bitnami_matomo
      - MYSQL_CLIENT_CREATE_DATABASE_USER=bn_matomo
      - MYSQL_CLIENT_CREATE_DATABASE_PASSWORD=bitnami
      - MATOMO_DATABASE_NAME=bitnami_matomo
      - MATOMO_DATABASE_USER=bn_matomo
      - MATOMO_DATABASE_PASSWORD=bitnami
      - MYSQL_AUTHENTICATION_PLUGIN=mysql_native_password
    volumes:
      - "matomo_data:/bitnami/matomo"
volumes:
  matomo_data:
    driver: local

```

Here's the docker logs for the above container:

```
matomo 00:25:07.58 
matomo 00:25:07.59 Welcome to the Bitnami matomo container
matomo 00:25:07.59 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-matomo
matomo 00:25:07.59 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-matomo/issues
matomo 00:25:07.59 
matomo 00:25:07.60 INFO  ==> ** Starting Matomo setup **
matomo 00:25:07.63 INFO  ==> Configuring PHP options
matomo 00:25:07.66 INFO  ==> Validating settings in MYSQL_CLIENT_* env vars
matomo 00:25:07.66 ERROR ==> The MARIADB_ROOT_PASSWORD environment variable is empty or not set. Set the environment variable ALLOW_EMPTY_PASSWORD=yes to allow the container to be started with blank passwords. This is recommended only for development.
```

Looks like there doesn't seem to be any environment variable specifically for defining "MYSQL" parameters, which confuses me as I am not sure if "MARIADB_*" are the right environment variables to connect to a MYSQL database server.


I have also tried creating a database and user manually in the in MYSQL first and then trying with the following docker-compose.yml file. Here's how I created the user:

```
CREATE USER 'bn_matomo'@'{host-machine-ip}' IDENTIFIED WITH mysql_native_password BY 'bitnami';

GRANT SELECT, INSERT, UPDATE, DELETE, CREATE, INDEX, DROP, ALTER, CREATE TEMPORARY TABLES, LOCK TABLES ON *.* TO 'bn_matomo'@'{host-machine-ip}';

GRANT FILE ON *.* TO 'bn_matomo'@'{host-machine-ip}';
```

```
version: '2'
services:
  matomo:
    image: "docker.io/bitnami/matomo:latest"
    user: root
    ports:
      - "84:8080"
    environment:
      - MARIADB_HOST={host-machine-ip}
      - MARIADB_PORT_NUMBER=3306
      - MARIADB_ROOT_USER={local-mysql-root-user}
      - MARIADB_ROOT_PASSWORD={local-mysql-root-password}
      - MATOMO_DATABASE_NAME=bitnami_matomo
      - MATOMO_DATABASE_USER=bn_matomo
      - MATOMO_DATABASE_PASSWORD=bitnami
      - MYSQL_AUTHENTICATION_PLUGIN=bitnami

    volumes:
      - "matomo_data:/bitnami/matomo"

volumes:
  matomo_data:
    driver: local

```

Here's the logs I see in that case

```
matomo 02:57:25.95 
matomo 02:57:25.95 Welcome to the Bitnami matomo container
matomo 02:57:25.96 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-matomo
matomo 02:57:25.96 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-matomo/issues
matomo 02:57:25.97 
matomo 02:57:25.97 INFO  ==> ** Starting Matomo setup **
matomo 02:57:26.05 INFO  ==> Configuring PHP options
matomo 02:57:26.07 INFO  ==> Validating settings in MYSQL_CLIENT_* env vars
matomo 02:57:26.18 INFO  ==> Persisted Matomo installation detected
matomo 02:57:26.18 INFO  ==> Updating default plugins in persisted data
matomo 02:57:26.30 INFO  ==> Restoring Matomo installation
realpath: /bitnami/matomo/config/config.ini.php: No such file or directory
```
 Could you please advice on this?

**Comment by marcosbc** [comment]
Hi @aayush27894, it seems like there is an issue with respect to the environment variable name that is said to be missing. You should be using the ones listed here (https://github.com/bitnami/bitnami-docker-matomo#create-a-database-for-matomo-using-mysql-client):

* `MYSQL_CLIENT_DATABASE_HOST`
* `MYSQL_CLIENT_DATABASE_PORT_NUMBER`
* `MYSQL_CLIENT_DATABASE_ROOT_USER`
* `MYSQL_CLIENT_DATABASE_ROOT_PASSWORD`
* `MYSQL_CLIENT_CREATE_DATABASE_NAME`
* `MYSQL_CLIENT_CREATE_DATABASE_USER`
* `MYSQL_CLIENT_CREATE_DATABASE_PASSWORD`
* `MYSQL_CLIENT_CREATE_DATABASE_CHARACTER_SET`
* `MYSQL_CLIENT_CREATE_DATABASE_COLLATE`

**Comment by marcosbc** [comment]
By the way, thanks for reporting this. We're fixing on a fix.

**Comment by aayush27894** [comment]
@marcosbc I did try using the above environment variables, but it seems to be giving me the same error as above. 

I believe that list of environment variables are used to create the database, user and password right?

Could you please tell me how you’d do it using the above suggested environment variables?
Could you please show what the docker compose file would look like if you were try to connect to a MySQL dB that is hosted on a local machine?

**Comment by marcosbc** [comment]
Hi @aayush27894, we just released a revision `4.2.1-debian-10-r35` which fixes any issues using `mysql-client`. Now if you use it with the environment variables above, it should work without issues. For example:

```yaml
  matomo:
    image: "docker.io/bitnami/matomo:latest"
    user: root
    ports:
      - "84:8080"
    environment:
      # Connection credentials for mysql-client
      - MYSQL_CLIENT_DATABASE_HOST=mariadbhost
      - MYSQL_CLIENT_DATABASE_PORT_NUMBER=3306
      - MYSQL_CLIENT_DATABASE_ROOT_USER=root
      - MYSQL_CLIENT_DATABASE_ROOT_PASSWORD=bitnami12345
      # Create database and a user to connect to it
      - MYSQL_CLIENT_CREATE_DATABASE_NAME=bitnami_matomo
      - MYSQL_CLIENT_CREATE_DATABASE_USER=bn_matomo
      - MYSQL_CLIENT_CREATE_DATABASE_PASSWORD=bitnami
      # Connection credentials for Matomo
      - MATOMO_DATABASE_HOST=mariadbhost
      - MATOMO_DATABASE_PORT_NUMBER=3306
      - MATOMO_DATABASE_NAME=bitnami_matomo
      - MATOMO_DATABASE_USER=bn_matomo
      - MATOMO_DATABASE_PASSWORD=bitnami
    volumes:
      - "matomo_data:/bitnami/matomo"
```

**Comment by aayush27894** [comment]
Hi @marcosbc, thanks for the support. This seemed to have worked.

However, this only works with MySQL 5.5 or older and that too only if I create the user and database manually first and then run the matomo docker-compose file. I am however, using MySQL (ver. 8.0.23).
I have tried using the following environment variable that you suggested:

```MYSQL_AUTHENTICATION_PLUGIN=mysql_native_password```

That however doesn't work for me. This is where is the docker logs is stuck:
```
matomo 21:54:35.44 
matomo 21:54:35.45 Welcome to the Bitnami matomo container
matomo 21:54:35.45 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-matomo
matomo 21:54:35.45 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-matomo/issues
matomo 21:54:35.46 
matomo 21:54:35.46 INFO  ==> ** Starting Matomo setup **
matomo 21:54:35.50 INFO  ==> Configuring PHP options
matomo 21:54:35.51 INFO  ==> Validating settings in MYSQL_CLIENT_* env vars
matomo 21:54:35.55 WARN  ==> You set the environment variable ALLOW_EMPTY_PASSWORD=yes. For safety reasons, do not use this flag in a production environment.
matomo 21:54:35.61 INFO  ==> Ensuring Matomo directories exist
matomo 21:54:35.62 INFO  ==> Trying to connect to the database server
```

Another issue I found was that when I used the create database and user environment variable. Here's how to reproduce it:
```
version: "2"

services:
  db:
    image: mysql:5.5
    restart: always
    ports:
      - "3308:3306"
    environment:
      - MYSQL_ROOT_PASSWORD=bitnami

```


```
version: '2'
services:
  matomo:
    image: "docker.io/bitnami/matomo:4.2.1-debian-10-r35"
    user: root
    ports:
      - "84:8080"
    environment:
      # Connection credentials for mysql-client
      - MYSQL_CLIENT_DATABASE_HOST={host}
      - MYSQL_CLIENT_DATABASE_PORT_NUMBER=3308
      - MYSQL_CLIENT_DATABASE_ROOT_USER=root
      - MYSQL_CLIENT_DATABASE_ROOT_PASSWORD=bitnami
      # Create database and a user to connect to it
      - MYSQL_CLIENT_CREATE_DATABASE_NAME=bitnami_matomo
      - MYSQL_CLIENT_CREATE_DATABASE_USER=bn_matomo
      - MYSQL_CLIENT_CREATE_DATABASE_PASSWORD=bitnami
      # Connection credentials for Matomo
      - MATOMO_DATABASE_HOST={host}
      - MATOMO_DATABASE_PORT_NUMBER=3308
      - MATOMO_DATABASE_NAME=bitnami_matomo
      - MATOMO_DATABASE_USER=bn_matomo
      - MATOMO_DATABASE_PASSWORD=bitnami
    volumes:
      - "mdata:/bitnami/matomo"

volumes:
  mdata:
    driver: local

```

It errors out here:

```
matomo 05:14:42.39 
matomo 05:14:42.39 Welcome to the Bitnami matomo container
matomo 05:14:42.40 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-matomo
matomo 05:14:42.40 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-matomo/issues
matomo 05:14:42.41 
matomo 05:14:42.41 INFO  ==> ** Starting Matomo setup **
matomo 05:14:42.48 INFO  ==> Configuring PHP options
matomo 05:14:42.50 INFO  ==> Validating settings in MYSQL_CLIENT_* env vars
matomo 05:14:42.50 INFO  ==> Trying to connect to the database server
matomo 05:14:42.52 INFO  ==> Creating database user bn_matomo
```

**Comment by marcosbc** [comment]
Hi @aayush27894, thanks for the details. You are indeed right again.

We have just released a new container revision, `4.2.1-debian-10-r37`, which adds support for a new [environment variable](https://github.com/bitnami/bitnami-docker-matomo/#create-a-database-for-matomo-using-mysql-client):

* `MYSQL_CLIENT_FLAVOR`: SQL database flavor. Valid values: `mariadb` or `mysql`. Default: **mariadb**.

Please re-deploy with the updated image, and specifying `MYSQL_CLIENT_FLAVOR=mysql`. I hope it works now!

**Comment by aayush27894** [comment]
This worked for me. This issue can be closed.
